### ISSUE TITLE
Windows installer fails with bundled JDK 17 (Cannot prepare bundled JVM to run the installer)

### ISSUE THREAD

**Issue by tonihele** [apache/netbeans#5783]
### Apache NetBeans version

Apache NetBeans 17

### What happened

We have a Netbeans platform project based on Netbeans 17 with bundled JDK 17 (Temurin). On Linux it seems to install correctly but on 64-bit Windows 11 & 10 the installer just fails with "Cannot prepare bundled JVM to run the installer".

I suspect this is due to us changing the bundled JDK from 11 to 17.

Our bug ticket:
https://github.com/jMonkeyEngine/sdk/issues/491

### How to reproduce

Those who are able, choose 64-bit Windows installer release.
https://github.com/jMonkeyEngine/sdk/releases/tag/v3.6-stable-sdk2

### Did this work correctly in an earlier version?

No / Don't know

### Operating System

Windows 11 & 10 64-bit

### JDK

jdk-17.0.6+10 Temurin; January 20, 2023; Windows x64

### Apache NetBeans packaging

Apache NetBeans platform

### Anything else
Extracted log:
[temp.txt](https://github.com/apache/netbeans/files/11184971/temp.txt)

_No response_

### Are you willing to submit a pull request?

No

**Comment by neilcsmith-net** [comment]
You could consider [NBPackage](https://github.com/apache/netbeans-nbpackage/), the Apache NetBeans packager, instead.  It's what we use for building the JDK included community installers of NetBeans at https://www.codelerity.com/netbeans/  I also use it (and wrote most of it) for platform usage.

**Comment by tonihele** [comment]
At this moment it seems like a big investment. I'm sure it is something very simple in the NB installer code. Just that I can't see any logs nor have any idea to get more information for some expert to fix this.

**Comment by tonihele** [comment]
Well, I was able to squeeze out a log file with:
`--verbose --output temp.txt`

The log lets me to believe pack200 is used and with JDK 17 it of course failed as pack200 was removed in JDK 14 already. This I believe is a setting we can set when building the installer bundle.

**Comment by neilcsmith-net** [comment]
@tonihele that could very well be the case, in which case you might be able to fix this issue at your end in the short term.  Is the Linux one you said works also bundling a JDK?

I don't think we really have "some expert" to fix NBI.  At least, there has been minimal development on it since the donation to ASF, and a bunch of outstanding bugs.  We've already stopped releasing NetBeans using the Linux NBI support due to unresolved issues.  Personally, I think we'll switch the Windows installer one way or another too.  And even if we don't, the JDK bundling can't be used at ASF so it's possibly even less likely to get looked at.

If you want some help with NBPackage, let me know - happy to help and feedback would be a welcomed contribution to it right now!  You can also see how it's currently integrated into the main NetBeans release builds (for DEB and RPM so far) here - https://github.com/apache/netbeans-jenkins-lib/blob/master/vars/asfMainNetBeansBuild.groovy#L394

**Comment by tonihele** [comment]
> @tonihele that could very well be the case, in which case you might be able to fix this issue at your end in the short term. Is the Linux one you said works also bundling a JDK?

Linux also is bundling JDK 17. So I don't exactly know why it works. Probably it finds other JDKs in the system and in Windows it doesn't then. We'll just try to disable the pack200, it looks like we have it enabled.

I'll probably create a ticket for us to try to move on to the NBPackage. But I don't see it happening anytime soon due to lack of manpower.

**Comment by tonihele** [comment]
Ok, I think this is a bug in the installer. I created a version without the pack200. Only to see that it fails too.

I think this happens because:
https://github.com/apache/netbeans-native-installers/blob/5cb81a27c543c92b1150c7c277d9cfd9d615e6b0/src/main/cpp/launcher/windows/src/JavaUtils.c#L638

Checks that does the pack200 exist. If not, everything fails.

I think it should perhaps not fail there immediately. I think it should fail in the `unpackJars` method IF it manages to find something to unpack to begin with.

**Comment by tonihele** [comment]
I was able to get this work for us by adding a fake `unpack200.exe` (0 bytes) to our bundled JDK 17. This then bypasses the check and as we don't have anything to unpack.... we are off the hook.

So yeah, I think this is a bug in NBI. Given pack200 was already removed in JDK 14... I think it is not fair to assume pack200 is there anymore.

**Comment by aaron-siegel** [comment]
Amazing. @tonihele THANK YOU for posting this. I ran into the exact same issue today, exactly 9 hours after you posted the fix! Creating a 0-byte `unpack200.exe` in the bundled JDK bin/ folder worked like a charm. Very fortunate timing... you've saved me countless hours of infuriating debugging.

@neilcsmith-net, or other NB folks, if you're listening - it is awfully silly for the installer to fail on a missing `unpack200.exe`, in cases where pack200 compression is disabled (so that `unpack200.exe` isn't actually necessary to complete the installation!)

More generally, there are evidently a number of us who rely on NBI to distribute our applications. It would be helpful to receive some official guidance from the NB team on the future of NBI. I've been using NetBeans since version 7, and I don't think the NBI code has changed much since then, if at all - despite being rife with typos and bugs. Is the use of installers still an officially supported workflow? Or is it effectively deprecated? If the latter, some documentation clarifying the supported alternatives, and detailing a suggested migration path, would be very welcome.

[Don't take this as criticism - NB is awesome! - more a desire for clarification, so that I can make informed decisions about the future architecture of my application.]

**Comment by neilcsmith-net** [comment]
No kidding.  There isn't any "official" guidance.  But in my opinion, it's dead Jim! :smile: It's not _actually_ deprecated, but certainly _effectively_ deprecated.  As you say, it's rife with bugs and no-one is looking at it.

So, yes, _an_ official option is NBPackage, in the sense it's been released by the Apache NetBeans project, although is still at beta stage.  Docs are just the readme so far, but I'm hoping to get around to more soon - it's on the to-do list!  In the meantime, happy to help.  Use GitHub discussions or the mailing lists for that though, please, rather than this issue.

https://github.com/apache/netbeans-nbpackage/
https://dlcdn.apache.org/netbeans/netbeans-nbpackage/1.0-beta2/

**Comment by ebarboni** [comment]
It's not totally correct 🛡️ 😄 . They are some fixes but never got released as we have some workflow build issue (egg chicken).
If you have a fix to Netbeans Installer github, we will be happy to merge it (I'm not c++/c) . So it may be available on NB18 a bit short but who knows.
I just call for a release vote for installer but I can cancel the vote.

Hope that nbpackage may help later but to be a complete replacement it should works nicely with nbm-maven-plugin way of handling installer artefact.

**Comment by tonihele** [comment]
I would imagine the bug is fairly easy to fix. I'm not a C person also. But the amount of code required is not really that big (I assume). Somebody who knows how to test it and all that, I'm sure it is very trivial to fix.

I mean one can just refactor it a bit:

- Function that returns files to be unpacked
- Function that unpacks them (requires & checks, only on this stage, the pack200)
- Now it is all crammed into one method, so essentially just a refactor

I'm sure it can be fixed in variety of ways. This is just one.

**Comment by neilcsmith-net** [comment]
> It's not totally correct shield smile . They are some fixes but never got released as we have some workflow build issue (egg chicken).

Fair point!  Sorry.  I was thinking more of issues and maintenance generally outside of the native layer (and our dropping of use for all but Windows IDE installers already).  If we can get a fix in for this issue in to NB18 that would be great.  The wider development issues can be discussed elsewhere.

**Comment by pedro-w** [comment]
I've just seen this bug today and I'm confident I can fix it. As I understand it the native part of the installer has been split out to https://github.com/apache/netbeans-native-installers by @ebarboni 
Maybe we can discuss this on the mailing list further (especially as to whether this is all going to be removed anyway)
It seems like at least JMonkeyEngine is still using it.

**Comment by mbien** [comment]
closing via https://github.com/apache/netbeans/issues/8281
